vespalib/src/vespa/vespalib/datastore/fixed_size_hash_map.cpp

```diff
@@ -0,0 +1,174 @@
+{
+    size_t hash_idx = comp.hash(EntryRef()) / _num_stripes;
+    hash_idx %= _modulo;
+    auto& chain_head = _chain_heads[hash_idx];
+    uint32_t node_idx = chain_head.load_relaxed();
+    while (node_idx != no_node_idx) {
+        auto& node = _nodes[node_idx];
+        if (comp.equal(EntryRef(), node.get_kv().first.load_relaxed())) {
+            return node.get_kv();
+        }
+        node_idx = node.get_next().load(std::memory_order_relaxed);
+    }
+    if (_free_head != no_node_idx) {
+        node_idx = _free_head;
+        auto& node = _nodes[node_idx];
+        _free_head = node.get_next().load(std::memory_order_relaxed);
+        --_free_count;
+        node.get_kv().first.store_release(insert_entry());
+        node.get_next().store(chain_head.load_relaxed());
+        chain_head.set(node_idx);
+        ++_count;
+        return node.get_kv();
+    }
+    assert(_nodes.size() < _nodes.capacity());
+    node_idx = _nodes.size();
+    new (_nodes.push_back_fast()) Node(std::make_pair(AtomicEntryRef(insert_entry()), AtomicEntryRef()), chain_head.load_relaxed());
+    chain_head.set(node_idx);
+    ++_count;
+    return _nodes[node_idx].get_kv();
+}
+
+void
+FixedSizeHashMap::transfer_hold_lists_slow(generation_t generation)
+{
+    auto &hold_2_list = _hold_2_list;
+    for (uint32_t node_idx : _hold_1_list) {
+        hold_2_list.push_back(std::make_pair(generation, node_idx));
+    }
+    _hold_1_list.clear();
+
+}
+
+
+void
+FixedSizeHashMap::trim_hold_lists_slow(generation_t usedGen)
+{
+    while (!_hold_2_list.empty()) {
+        auto& first = _hold_2_list.front();
+        if (static_cast<sgeneration_t>(first.first - usedGen) >= 0) {
+            break;
+        }
+        uint32_t node_idx = first.second;
+        auto& node = _nodes[node_idx];
+        node.get_next().store(_free_head, std::memory_order_relaxed);
+        _free_head = node_idx;
+        ++_free_count;
+        --_hold_count;
+        node.on_free();
+        _hold_2_list.erase(_hold_2_list.begin());
+    }
+}
+
+FixedSizeHashMap::KvType*
+FixedSizeHashMap::remove(const EntryComparator& comp, EntryRef key_ref)
+{
+    size_t hash_idx = comp.hash(key_ref) / _num_stripes;
+    hash_idx %= _modulo;
+    auto& chain_head = _chain_heads[hash_idx];
+    uint32_t node_idx = chain_head.load_relaxed();
+    uint32_t prev_node_idx = no_node_idx;
+    while (node_idx != no_node_idx) {
+        auto &node = _nodes[node_idx];
+        uint32_t next = node.get_next().load(std::memory_order_relaxed);
+        if (comp.equal(key_ref, node.get_kv().first.load_relaxed())) {
+            if (prev_node_idx != no_node_idx) {
+                _nodes[prev_node_idx].get_next().store(next, std::memory_order_release);
+            } else {
+                chain_head.set(next);
+            }
+            --_count;
+            ++_hold_count;
+            _hold_1_list.push_back(node_idx);
+            return &_nodes[node_idx].get_kv();
+        }
+        node_idx = next;
```

Comment:
I think _prev_node_idx_ should be updated here as well. Otherwise the removed element cannot be removed from the node link chain as L140 never will trigger. I also think we should write separate unit tests for FixedSizeHashMap, such that all edge cases and intricacies are tested explicitly.